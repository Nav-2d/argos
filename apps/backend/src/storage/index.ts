export * from "./checkIfExists.js";
export * from "./download.js";
export * from "./get.js";
export * from "./getPublicUrl.js";
export * from "./upload.js";
export * from "./signed-url.js";
export * from "./ImageFile.js";
export * from "./s3.js";
