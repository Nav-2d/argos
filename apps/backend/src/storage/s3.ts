import { S3Client } from "@aws-sdk/client-s3";

let client: S3Client;

export type { S3Client } from "@aws-sdk/client-s3";

export const getS3Client = () => {
  if (!client) {
    client = new S3Client({ region: "eu-west-1" });
  }
  return client;
};
