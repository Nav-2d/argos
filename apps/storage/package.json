{
  "name": "@argos-ci/storage",
  "version": "2.0.0",
  "private": true,
  "exports": "./dist/index.js",
  "type": "module",
  "dependencies": {
    "@argos-ci/config": "*",
    "@aws-sdk/client-s3": "^3.386.0",
    "@aws-sdk/s3-request-presigner": "^3.386.0",
    "mime": "^3.0.0",
    "sharp": "^0.32.4",
    "tmp": "^0.2.1"
  },
  "devDependencies": {
    "@types/mime": "^3.0.1",
    "@types/sharp": "^0.31.1",
    "@types/tmp": "^0.2.3",
    "axios": "^1.4.0"
  },
  "scripts": {
    "build": "swc src --config-file ../../.swcrc -d dist",
    "watch-build": "npm run build -- --watch --quiet"
  }
}
