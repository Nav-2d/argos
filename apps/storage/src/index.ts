export * from "./checkIfExists.js";
export * from "./download.js";
export * from "./get.js";
export * from "./upload.js";
export * from "./signed-url.js";
export * from "./s3.js";
