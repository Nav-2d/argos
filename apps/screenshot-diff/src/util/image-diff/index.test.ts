import { join } from "node:path";
import { fileURLToPath } from "node:url";

import { S3Image, s3 as s3client } from "@argos-ci/storage";

import { diffImages } from "./index.js";

const __dirname = fileURLToPath(new URL(".", import.meta.url));
const s3 = s3client();

describe("#diffImages", () => {
  it("simple", async () => {
    const result = await diffImages({
      baseImage: new S3Image({
        s3,
        filePath: join(__dirname, "__fixtures__/simple/compare.png"),
        localFile: true,
        protectOriginal: true,
      }),
      compareImage: new S3Image({
        s3,
        filePath: join(__dirname, "__fixtures__/simple/base.png"),
        localFile: true,
        protectOriginal: true,
      }),
      diffImage: new S3Image({
        s3,
        filePath: join(__dirname, "__fixtures__/simple/diff_tmp.png"),
        localFile: true,
        protectOriginal: true,
      }),

      fuzz: 900,
    });

    expect(result.score).toBeCloseTo(0.306, 2);
    expect(result.pixels).toBeCloseTo(50e4, -4);
  });

  it("simple with enough fuzz", async () => {
    const result = await diffImages({
      baseImage: new S3Image({
        s3,
        filePath: join(__dirname, "__fixtures__/simple/compare.png"),
        localFile: true,
        protectOriginal: true,
      }),
      compareImage: new S3Image({
        s3,
        filePath: join(__dirname, "__fixtures__/simple/base.png"),
        localFile: true,
        protectOriginal: true,
      }),
      diffImage: new S3Image({
        s3,
        filePath: join(__dirname, "__fixtures__/simple/diff_tmp.png"),
        localFile: true,
        protectOriginal: true,
      }),
      fuzz: 70 ** 2,
    });

    expect(result.score).toBe(0);
    expect(result).toMatchSnapshot();
  });

  it("alphaBackground", async () => {
    const result = await diffImages({
      baseImage: new S3Image({
        s3,
        filePath: join(__dirname, "__fixtures__/alphaBackground/compare.png"),
        localFile: true,
        protectOriginal: true,
      }),
      compareImage: new S3Image({
        s3,
        filePath: join(__dirname, "__fixtures__/alphaBackground/base.png"),
        localFile: true,
        protectOriginal: true,
      }),
      diffImage: new S3Image({
        s3,
        filePath: join(__dirname, "__fixtures__/alphaBackground/diff_tmp.png"),
        localFile: true,
        protectOriginal: true,
      }),
    });

    expect(result.score).toBe(0);
    expect(result).toMatchSnapshot();
  });

  it("boxShadow", async () => {
    const result = await diffImages({
      baseImage: new S3Image({
        s3,
        filePath: join(__dirname, "__fixtures__/boxShadow/compare.png"),
        localFile: true,
        protectOriginal: true,
      }),
      compareImage: new S3Image({
        s3,
        filePath: join(__dirname, "__fixtures__/boxShadow/base.png"),
        localFile: true,
        protectOriginal: true,
      }),
      diffImage: new S3Image({
        s3,
        filePath: join(__dirname, "__fixtures__/boxShadow/diff_tmp.png"),
        localFile: true,
        protectOriginal: true,
      }),
    });

    expect(result.score).toBe(0);
    expect(result).toMatchSnapshot();
  });

  it("border", async () => {
    const result = await diffImages({
      baseImage: new S3Image({
        s3,
        filePath: join(__dirname, "__fixtures__/border/compare.png"),
        localFile: true,
        protectOriginal: true,
      }),
      compareImage: new S3Image({
        s3,
        filePath: join(__dirname, "__fixtures__/border/base.png"),
        localFile: true,
        protectOriginal: true,
      }),
      diffImage: new S3Image({
        s3,
        filePath: join(__dirname, "__fixtures__/border/diff_tmp.png"),
        localFile: true,
        protectOriginal: true,
      }),
    });

    expect(result.score).toBe(0);
  });

  it("fontAliasing", async () => {
    const result = await diffImages({
      baseImage: new S3Image({
        s3,
        filePath: join(__dirname, "__fixtures__/fontAliasing/compare.png"),
        localFile: true,
        protectOriginal: true,
      }),
      compareImage: new S3Image({
        s3,
        filePath: join(__dirname, "__fixtures__/fontAliasing/base.png"),
        localFile: true,
        protectOriginal: true,
      }),
      diffImage: new S3Image({
        s3,
        filePath: join(__dirname, "__fixtures__/fontAliasing/diff_tmp.png"),
        localFile: true,
        protectOriginal: true,
      }),
    });

    expect(result.score).toBeCloseTo(0, 2);
  });

  it("imageCompression", async () => {
    const result = await diffImages({
      baseImage: new S3Image({
        s3,
        filePath: join(__dirname, "__fixtures__/imageCompression/compare.png"),
        localFile: true,
        protectOriginal: true,
      }),
      compareImage: new S3Image({
        s3,
        filePath: join(__dirname, "__fixtures__/imageCompression/base.png"),
        localFile: true,
        protectOriginal: true,
      }),
      diffImage: new S3Image({
        s3,
        filePath: join(__dirname, "__fixtures__/imageCompression/diff_tmp.png"),
        localFile: true,
        protectOriginal: true,
      }),
    });

    expect(result.score).toBe(0);
    expect(result).toMatchSnapshot();
  });

  it("imageCompression2", async () => {
    const result = await diffImages({
      baseImage: new S3Image({
        s3,
        filePath: join(__dirname, "__fixtures__/imageCompression2/compare.png"),
        localFile: true,
        protectOriginal: true,
      }),
      compareImage: new S3Image({
        s3,
        filePath: join(__dirname, "__fixtures__/imageCompression2/base.png"),
        localFile: true,
        protectOriginal: true,
      }),
      diffImage: new S3Image({
        s3,
        filePath: join(
          __dirname,
          "__fixtures__/imageCompression2/diff_tmp.png"
        ),
        localFile: true,
        protectOriginal: true,
      }),
    });

    expect(result.score).toBe(0);
    expect(result.scoreRaw).toBeLessThan(1e-5);
  });

  it("imageCompression3", async () => {
    const result = await diffImages({
      baseImage: new S3Image({
        s3,
        filePath: join(__dirname, "__fixtures__/imageCompression3/compare.png"),
        localFile: true,
        protectOriginal: true,
      }),
      compareImage: new S3Image({
        s3,
        filePath: join(__dirname, "__fixtures__/imageCompression3/base.png"),
        localFile: true,
        protectOriginal: true,
      }),
      diffImage: new S3Image({
        s3,
        filePath: join(
          __dirname,
          "__fixtures__/imageCompression3/diff_tmp.png"
        ),
        localFile: true,
        protectOriginal: true,
      }),
    });

    expect(result.pixels).toBeCloseTo(35, -2);
    expect(result.score).toBeCloseTo(0, 3);
  });

  it("big images", async () => {
    const result = await diffImages({
      baseImage: new S3Image({
        s3,
        filePath: join(__dirname, "__fixtures__/big-images/compare.png"),
        localFile: true,
        protectOriginal: true,
      }),
      compareImage: new S3Image({
        s3,
        filePath: join(__dirname, "__fixtures__/big-images/base.png"),
        localFile: true,
        protectOriginal: true,
      }),
      diffImage: new S3Image({
        s3,
        filePath: join(__dirname, "__fixtures__/big-images/diff_tmp.png"),
        localFile: true,
        protectOriginal: true,
      }),
    });

    expect(result.score).toBeGreaterThan(0.95);
  });
});
