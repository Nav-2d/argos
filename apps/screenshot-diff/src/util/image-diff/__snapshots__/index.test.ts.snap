// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`#diffImages alphaBackground 1`] = `
{
  "height": 300,
  "pixels": 0,
  "score": 0,
  "scoreRaw": 0,
  "width": 1400,
}
`;

exports[`#diffImages boxShadow 1`] = `
{
  "height": 300,
  "pixels": 0,
  "score": 0,
  "scoreRaw": 0,
  "width": 250,
}
`;

exports[`#diffImages imageCompression 1`] = `
{
  "height": 665,
  "pixels": 0,
  "score": 0,
  "scoreRaw": 0,
  "width": 327,
}
`;

exports[`#diffImages simple with enough fuzz 1`] = `
{
  "height": 1146,
  "pixels": 0,
  "score": 0,
  "scoreRaw": 0,
  "width": 1425,
}
`;
